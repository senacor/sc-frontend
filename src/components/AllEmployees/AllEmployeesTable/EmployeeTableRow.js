import React, { useState, Fragment } from 'react';
import { injectIntl } from 'react-intl';
import { withStyles } from '@material-ui/core';
import {
  formatLocaleDateTime,
  FRONTEND_DATE_FORMAT
} from '../../../helper/date';

// Material UI
import TableRow from '@material-ui/core/TableRow';
import TableCell from '@material-ui/core/TableCell';
import EmployeesPRsDialog from '../EmployeesPRsDialog';

const styles = theme => ({
  tableRow: {
    height: 80,
    '&:hover': {
      transition: 'all 0.3s'
    }
  },
  notSelection: {
    cursor: 'pointer',
    '&:hover': {
      backgroundColor: theme.palette.secondary.brighterGrey
    }
  },
  selectionUnavailable: {
    backgroundColor: theme.palette.secondary.grey,
    cursor: 'auto'
  },
  selected: {
    backgroundColor: theme.palette.secondary.green
  },
  selectable: {
    cursor: 'cell'
  }
});

const EmployeeTableRow = ({
  classes,
  employee: {
    id,
    firstName,
    lastName,
    competenceCenter,
    currentPosition,
    currentCst,
    officeLocation,
    dateOfNextPr,
    supervisorName,
    hasOpenedPr
  },
  selection,
  selected,
  toggleSelected
}) => {
  const [dialogOpen, setDialogOpen] = useState(false);

  const handleDialogOpen = () => {
    setDialogOpen(true);
  };

  const handleDialogClose = () => {
    setDialogOpen(false);
  };

  const employeeName = `${firstName} ${lastName}`;
  let bgClass = '';
  let onRowClick = handleDialogOpen;

  if (selection) {
    if (hasOpenedPr || !supervisorName) {
      bgClass = classes.selectionUnavailable;
      onRowClick = () => {};
    } else {
      onRowClick = () => toggleSelected(id);
      bgClass = selected ? classes.selected : classes.selectable;
    }
  } else {
    bgClass = classes.notSelection;
  }

  return (
    <Fragment>
      <TableRow
        className={`${classes.tableRow} ${bgClass}`}
        onClick={onRowClick}
      >
        <TableCell>{employeeName}</TableCell>
        <TableCell>{currentPosition}</TableCell>
        <TableCell>{currentCst}</TableCell>
        <TableCell>{supervisorName}</TableCell>
        <TableCell>{competenceCenter}</TableCell>
        <TableCell>{officeLocation}</TableCell>
        <TableCell>
          {formatLocaleDateTime(dateOfNextPr, FRONTEND_DATE_FORMAT)}
        </TableCell>
      </TableRow>
      {dialogOpen && (
        <EmployeesPRsDialog
          firstName={firstName}
          lastName={lastName}
          employeeId={id}
          dialogOpen={dialogOpen}
          dialogClose={handleDialogClose}
        />
      )}
    </Fragment>
  );
};

export default injectIntl(withStyles(styles)(EmployeeTableRow));
