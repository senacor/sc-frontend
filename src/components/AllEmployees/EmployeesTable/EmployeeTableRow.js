import React, { Fragment, useState } from 'react';
import { injectIntl } from 'react-intl';
import { Typography, withStyles } from '@material-ui/core';
import {
  formatLocaleDateTime,
  FRONTEND_DATE_FORMAT
} from '../../../helper/date';
// Material UI
import TableRow from '@material-ui/core/TableRow';
import TableCell from '@material-ui/core/TableCell';
import { translateGeneralStatus } from '../../../helper/string';
import EmployeeFilter from '../../admin/EmployeeFilter';
import { changeSupervisor } from '../../../calls/employees';
import ROUTES from '../../../helper/routes';

const styles = theme => ({
  tableRow: {
    height: 80,
    '&:hover': {
      transition: 'all 0.3s'
    }
  },
  notSelection: {
    cursor: 'pointer',
    '&:hover': {
      backgroundColor: theme.palette.secondary.brighterGrey
    }
  },
  selectionUnavailable: {
    backgroundColor: theme.palette.secondary.grey,
    cursor: 'auto'
  },
  selected: {
    backgroundColor: theme.palette.secondary.green
  },
  selectable: {
    cursor: 'cell'
  }
});

const EmployeeTableRow = ({
  classes,
  intl,
  employee: {
    id,
    firstName,
    lastName,
    department,
    position,
    scStatus,
    officeLocation,
    supervisorName,
    entryDate,
    endDate
  },
  formerEmployee,
  user,
  info,
  error,
  currentSupervisors,
  setSelectedEmployee
}) => {
  const [supervisorNameState, setSupervisorNameState] = useState(
    supervisorName
  );

  const handleDialogOpen = () => {
    window.history.pushState(null, null, `${ROUTES.EMPLOYEE_SC}/${id}`);
    setSelectedEmployee({
      id: id,
      firstName: firstName,
      lastName: lastName,
      supervisorName: supervisorName
    });
  };

  const employeeName = `${firstName} ${lastName}`;

  const handleChangeSupervisor = supervisor => {
    changeSupervisor(id, supervisor.id, info, error).then(() => {
      setSupervisorNameState(`${supervisor.firstName} ${supervisor.lastName}`);
    });
  };

  return (
    <Fragment>
      <TableRow
        className={`${classes.tableRow} ${classes.notSelection}`}
        onClick={handleDialogOpen}
      >
        <TableCell>{employeeName}</TableCell>
        <TableCell>{position}</TableCell>
        {!formerEmployee && (
          <TableCell>
            {scStatus
              ? intl.formatMessage({ id: translateGeneralStatus(scStatus) })
              : intl.formatMessage({ id: 'employeeInfo.noScStatus' })}
          </TableCell>
        )}
        <TableCell>
          {!formerEmployee && (
            <Typography>
              {user.hasRoleHr() ? (
                <EmployeeFilter
                  data={currentSupervisors}
                  supervisorName={
                    supervisorNameState
                      ? supervisorNameState
                      : intl.formatMessage({
                          id: 'employeeInfo.noSupervisor'
                        })
                  }
                  setSelectedEmployee={handleChangeSupervisor}
                />
              ) : (
                <TableCell>
                  <span className={classes.textInfo}>
                    {supervisorName
                      ? supervisorName
                      : intl.formatMessage({ id: 'employeeInfo.noSupervisor' })}
                  </span>
                </TableCell>
              )}
            </Typography>
          )}
        </TableCell>
        <TableCell>{department}</TableCell>
        <TableCell>{officeLocation}</TableCell>
        <TableCell>
          {formerEmployee
            ? formatLocaleDateTime(endDate, FRONTEND_DATE_FORMAT)
            : formatLocaleDateTime(entryDate, FRONTEND_DATE_FORMAT)}
        </TableCell>
      </TableRow>
    </Fragment>
  );
};

export default injectIntl(withStyles(styles)(EmployeeTableRow));
