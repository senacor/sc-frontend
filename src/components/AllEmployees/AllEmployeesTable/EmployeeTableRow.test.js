import React from 'react';
import { shallowWithIntl } from 'enzyme-react-intl';

import EmployeeTableRow from './EmployeeTableRow';

let wrapper;
const mockSelection = false;
const mockEmployee = {
  id: 1,
  firstName: 'Peter',
  lastName: 'Knox',
  competenceCenter: 'DEVELOPMENT',
  currentPosition: 'Senior Developer',
  currentCst: 'Postbank',
  officeLocation: 'KoÅ¡ice',
  dateOfNextPr: [1, 2, 2020],
  userPhoto: undefined,
  supervisorName: 'Peter Supervisor',
  hasOpenedPr: false
};

describe('EmployeeTableRow', () => {
  wrapper = shallowWithIntl(
    <EmployeeTableRow selection={mockSelection} employee={mockEmployee} />
  );
  wrapper.setState({ dialogOpen: false });
  it('tests selection prop to be false', () => {
    expect(wrapper.props().selection).toEqual(false);
  });
  it('tests employee prop to be included', () => {
    expect(wrapper.props().employee).toEqual(mockEmployee);
  });
  it('checks for employee name', () => {
    expect(wrapper.props().employee.firstName).toBeTruthy();
    expect(wrapper.props().employee.lastName).toBeTruthy();
  });
});
