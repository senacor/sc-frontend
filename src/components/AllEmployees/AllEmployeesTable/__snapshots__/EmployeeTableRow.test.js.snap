// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeeTableRow should render correctly 1`] = `
<WithStyles(EmployeeTableRow)
  employee={
    Object {
      "competenceCenter": "DEVELOPMENT",
      "currentCst": "Postbank",
      "currentPosition": "Senior Developer",
      "dateOfNextPr": Array [
        1,
        2,
        2020,
      ],
      "firstName": "Peter",
      "hasOpenedPr": false,
      "id": 1,
      "lastName": "Knox",
      "officeLocation": "KoÅ¡ice",
      "supervisorName": "Peter Supervisor",
      "userPhoto": undefined,
    }
  }
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
  selection={false}
/>
`;
